File: Data Structures & Algorithms/eating-bananas/submission-2.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        // Take values of potential k from 1 to max(piles) in and array
        // apply binary serach in the array of k values.
        // for every k value count the number of hours to complete all the piles.
        // if  number of hours to complete all the piles <= h , Take res = min(res, k)
        // move the right pointer to k-1.
        // else move the left pointer to k + 1.
        

        int l = 1;
        int r = Arrays.stream(piles).max().getAsInt();
        int k, totalHours;
        int res = r;
        
        while(l<=r) {
            totalHours = 0;
            k = (int)((l + r) / 2); 
            for(int p : piles){
                totalHours += Math.ceil((double) p / k);
            }
            if(totalHours <= h) {
                res = Math.min(res, k);
                r = k - 1;
            } else {
                l = k + 1;
            }
        }
        return res;   
    }
}

File: Data Structures & Algorithms/max-water-container/submission-2.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        max_amount = 0

        left = 0
        right = len(heights) - 1
        while (left < right):
            if(heights[left] <= heights[right]):
                length = heights[left]
            else:
                length = heights[right]

            amount = length * (right - left)
            max_amount = max(max_amount , amount)
            
            if (heights[left] < heights[right]):
                left += 1
            else:
                right -= 1

        return max_amount
